File: app/schemas/broadcast.py
from typing import Any

from pydantic import BaseModel, Field

# Mirrors the CHECK constraint in migrations/005_broadcast_interval_floor.sql.
MIN_INTERVAL_SECONDS = 1
MAX_INTERVAL_SECONDS = 60
DEFAULT_INTERVAL_SECONDS = 2

# Round gap for looping tasks (migrations/009_broadcast_loop.sql). Capped so a
# typo cannot park a task for months, and floored because a round already takes
# total_count * interval_seconds to deliver.
MIN_LOOP_INTERVAL_SECONDS = 5
MAX_LOOP_INTERVAL_SECONDS = 86400
DEFAULT_LOOP_INTERVAL_SECONDS = 60
MAX_LOOP_TOTAL = 50


class BroadcastTaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    group_ids: list[str] = Field(min_length=1)
    message: list[dict[str, Any]] = Field(min_length=1)
    interval_seconds: int = Field(
        default=DEFAULT_INTERVAL_SECONDS, ge=MIN_INTERVAL_SECONDS, le=MAX_INTERVAL_SECONDS
    )
    group_cooldown_seconds: int = Field(default=0, ge=0, le=86400)
    # 0 (the default) sends the schedule once; 1 is equivalent to 0.
    loop_total: int = Field(default=0, ge=0, le=MAX_LOOP_TOTAL)
    loop_interval_seconds: int = Field(
        default=DEFAULT_LOOP_INTERVAL_SECONDS,
        ge=MIN_LOOP_INTERVAL_SECONDS,
        le=MAX_LOOP_INTERVAL_SECONDS,
    )


class BroadcastResumePayload(BaseModel):
    """Optional new group delay applied while resuming a paused task."""

    interval_seconds: int | None = Field(
        default=None, ge=MIN_INTERVAL_SECONDS, le=MAX_INTERVAL_SECONDS
    )


class BroadcastIntervalsPayload(BaseModel):
    """Retune a task's pacing without touching its schedule.

    ``loop_interval_seconds`` only decides when the *next* round starts, so it is
    safe to change at any time. ``interval_seconds`` re-spaces the groups that
    are still queued, so it stays restricted to a paused task — the same rule
    the resume route already applies.
    """

    interval_seconds: int | None = Field(
        default=None, ge=MIN_INTERVAL_SECONDS, le=MAX_INTERVAL_SECONDS
    )
    loop_interval_seconds: int | None = Field(
        default=None, ge=MIN_LOOP_INTERVAL_SECONDS, le=MAX_LOOP_INTERVAL_SECONDS
    )

File: web/src/api/broadcast.ts
import type { BroadcastTask, BroadcastTaskDetail, MessageSegment } from "../types/api";
import { apiJson } from "./client";

export function listBroadcastTasks(limit?: number): Promise<BroadcastTask[]> {
  return apiJson<BroadcastTask[]>(`/api/broadcast-tasks${limit ? `?limit=${limit}` : ""}`);
}

export function getBroadcastTask(id: string): Promise<BroadcastTaskDetail> {
  return apiJson<BroadcastTaskDetail>(`/api/broadcast-tasks/${id}`);
}

export function createBroadcastTask(payload: {
  title: string;
  group_ids: string[];
  message: MessageSegment[];
  interval_seconds: number;
  /** 0 sends the schedule once. */
  loop_total: number;
  loop_interval_seconds: number;
}): Promise<unknown> {
  return apiJson("/api/broadcast-tasks", { method: "POST", body: JSON.stringify(payload) });
}

/**
 * Retune pacing. The round gap is accepted at any time; `interval_seconds`
 * re-spaces already queued sends, so the backend only takes it while paused.
 */
export function updateBroadcastIntervals(
  id: string,
  payload: { interval_seconds?: number; loop_interval_seconds?: number },
): Promise<unknown> {
  return apiJson(`/api/broadcast-tasks/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function pauseBroadcastTask(id: string): Promise<unknown> {
  return apiJson(`/api/broadcast-tasks/${id}/pause`, { method: "POST" });
}

/** Resume a paused task, optionally re-tuning the group delay in the same call. */
export function resumeBroadcastTask(id: string, intervalSeconds?: number): Promise<unknown> {
  return apiJson(`/api/broadcast-tasks/${id}/resume`, {
    method: "POST",
    body: JSON.stringify({ interval_seconds: intervalSeconds ?? null }),
  });
}

export function cancelBroadcastTask(id: string): Promise<unknown> {
  return apiJson(`/api/broadcast-tasks/${id}/cancel`, { method: "POST" });
}